File: webapp/frontend/src/commandGenerator.js
/**
 * Command taxonomy + slot-filled templates + debouncing.
 *
 * Direct port of learning_runtime/command_generator.py. Command categories:
 * orient, confirm, correct_tool, correct_motion, correct_sequence,
 * encourage, safety. Templates are pulled from the active step's
 * commandTemplates field, authored offline (see
 * scripts/generate_step_templates.py) and reviewed by a human. If that
 * field is empty or still unreviewed, this falls back to a generic
 * templated sentence built from narration and logs a console warning — it
 * never fails silently, and it never delivers an unreviewed LLM-drafted
 * command as if it were human-approved coaching copy.
 */

// Mirrors scripts/generate_step_templates.py's COMMAND_CATEGORIES (and
// learning_runtime/command_generator.py's copy of the same list). Kept as a
// separate constant, not imported — this module has no dependency on the
// Python pipeline.
export const COMMAND_CATEGORIES = [
  "orient",
  "confirm",
  "correct_tool",
  "correct_motion",
  "correct_sequence",
  "encourage",
  "safety",
];

const STATE_TO_CATEGORY = {
  on_track: "encourage",
  wrong_tool: "correct_tool",
  wrong_hand_state: "correct_motion",
  wrong_sequence: "correct_sequence",
  step_complete: "confirm",
  no_evidence: "orient",
};

const GENERIC_FALLBACKS = {
  orient: (narration) => `Get your hands back on the workbench for this step: ${narration}`,
  confirm: (narration) => `That matches this step — nice work: ${narration}`,
  correct_tool: (narration) => `Try picking up the tool for this step: ${narration}`,
  correct_motion: (narration) => `Adjust your hand placement for this step: ${narration}`,
  correct_sequence: (narration) => `Hold on — that's a different step. The current step is: ${narration}`,
  encourage: (narration) => `Good, keep going: ${narration}`,
  safety: (narration) => `Take care here: ${narration}`,
};

export class CommandGenerator {
  constructor({ debounceSeconds = 8.0, clock = () => performance.now() / 1000 } = {}) {
    this.debounceSeconds = debounceSeconds;
    this._clock = clock;
    this._lastEmitted = new Map(); // `${stepIndex}:${category}` -> last-emitted time
  }

  generate(comparison, step) {
    if (comparison.state === "no_evidence") return null; // nothing actionable to say about silence

    const category = STATE_TO_CATEGORY[comparison.state];
    const key = `${step.stepIndex}:${category}`;
    const now = this._clock();
    const last = this._lastEmitted.get(key);
    if (last !== undefined && now - last < this.debounceSeconds) {
      return null; // same correction for this step was just given — don't repeat every frame
    }

    this._lastEmitted.set(key, now);
    return { category, text: this._selectTemplate(category, step), stepIndex: step.stepIndex };
  }

  _selectTemplate(category, step) {
    const templates = step.commandTemplatesReviewed ? step.commandTemplates?.[category] : null;
    if (templates && templates.length > 0) {
      return templates[0]
        .replaceAll("{narration}", step.narration)
        .replaceAll("{required_tools}", (step.requiredTools ?? []).join(", "))
        .replaceAll("{gaze_focus}", step.gazeFocus ?? "");
    }
    console.warn(
      `step ${step.stepIndex}: command_templates missing or unreviewed for category '${category}' — ` +
        "falling back to a generic narration-based sentence instead of failing silently"
    );
    return GENERIC_FALLBACKS[category](step.narration);
  }
}

File: learning_runtime/command_generator.py
"""Command taxonomy + slot-filled templates + debouncing.

Command categories: orient, confirm, correct_tool, correct_motion,
correct_sequence, encourage, safety. Templates are pulled from the active
step's command_templates field, authored offline (see
scripts/generate_step_templates.py) and reviewed by a human. If that field is
empty or still unreviewed, this falls back to a generic templated sentence
built from narration_text and logs a warning — it never fails silently, and
it never delivers an unreviewed LLM-drafted command as if it were
human-approved coaching copy.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from typing import Callable

from .comparator import ComparisonResult, ComparisonState
from .step_templates import StepTemplate

logger = logging.getLogger(__name__)

# Mirrors scripts/generate_step_templates.py's COMMAND_CATEGORIES. Kept as a
# separate constant (not imported) — learning_runtime has no dependency on
# the offline authoring pipeline.
COMMAND_CATEGORIES = (
    "orient",
    "confirm",
    "correct_tool",
    "correct_motion",
    "correct_sequence",
    "encourage",
    "safety",
)

_STATE_TO_CATEGORY = {
    ComparisonState.ON_TRACK: "encourage",
    ComparisonState.WRONG_TOOL: "correct_tool",
    ComparisonState.WRONG_HAND_STATE: "correct_motion",
    ComparisonState.WRONG_SEQUENCE: "correct_sequence",
    ComparisonState.STEP_COMPLETE: "confirm",
    ComparisonState.NO_EVIDENCE: "orient",
}

_GENERIC_FALLBACKS = {
    "orient": "Get your hands back on the workbench for this step: {narration}",
    "confirm": "That matches this step — nice work: {narration}",
    "correct_tool": "Try picking up the tool for this step: {narration}",
    "correct_motion": "Adjust your hand placement for this step: {narration}",
    "correct_sequence": "Hold on — that's a different step. The current step is: {narration}",
    "encourage": "Good, keep going: {narration}",
    "safety": "Take care here: {narration}",
}


@dataclass
class Command:
    category: str
    text: str
    step_index: int


class CommandGenerator:
    def __init__(self, debounce_seconds: float = 8.0, clock: Callable[[], float] = time.monotonic):
        self.debounce_seconds = debounce_seconds
        self._clock = clock
        self._last_emitted: dict[tuple[int, str], float] = {}  # (step_index, category) -> last-emitted time

    def generate(self, comparison: ComparisonResult, step: StepTemplate) -> Command | None:
        if comparison.state == ComparisonState.NO_EVIDENCE:
            return None  # nothing actionable to say about silence

        category = _STATE_TO_CATEGORY[comparison.state]
        key = (step.step_index, category)
        now = self._clock()
        last = self._last_emitted.get(key)
        if last is not None and (now - last) < self.debounce_seconds:
            return None  # same correction for this step was just given — don't repeat every frame

        self._last_emitted[key] = now
        return Command(category=category, text=self._select_template(category, step), step_index=step.step_index)

    def _select_template(self, category: str, step: StepTemplate) -> str:
        templates = step.command_templates.get(category) if step.command_templates_reviewed else None
        if templates:
            return templates[0].format(
                narration=step.narration,
                required_tools=", ".join(step.required_tools),
                gaze_focus=step.gaze_focus or "",
            )
        logger.warning(
            "step %s: command_templates missing or unreviewed for category %r — falling back to a generic "
            "narration-based sentence instead of failing silently",
            step.step_index,
            category,
        )
        return _GENERIC_FALLBACKS[category].format(narration=step.narration)
